### PR TITLE
Add Usage Insights canvas extension 🤖🤖🤖

### PR DESCRIPTION
**Pull request by davidkaya**
## Description

Adds the **Usage Insights** canvas as a reusable canvas extension plus a matching installable plugin.

The canvas provides current and historical GitHub Copilot token and AI-credit metrics, per-agent and sub-agent rollups, model-call details, and in-canvas history drill-down. It reads local Copilot usage data in read-only mode and does not add external session deep links.

The preview below is rendered from the real canvas using entirely synthetic sample metrics, generic session names, generic agent names, and no local user data.

Usage Insights canvas showing synthetic sample metrics

## What's included

- `extensions/usage-insights/` — pinned extension functionality from `davidkaya/github-copilot-usage-insights@206d53faab1d8a3331de96e203d0206ad21845aa`, adapted to the `usage-insights` ID and exact **Usage Insights** display name.
- `plugins/usage-insights/` — Agent Plugins v1.0.0 manifest registering `./extensions/usage-insights`, plus installation and privacy documentation.
- `extensions/usage-insights/assets/preview.png` — polished desktop preview generated with synthetic, sanitized sample data.
- Regenerated `.github/plugin/marketplace.json`, `docs/README.plugins.md`, and website data.

## Checklist

- [x] `npm ci --registry=https://packagefeedproxy.microsoft.io/npm` passes
- [x] `npm run plugin:validate` passes (`usage-insights` valid)
- [x] `npm run build` regenerates marketplace + docs
- [x] `npm run website:data` regenerated
- [x] Extension syntax and composition paths verified
- [x] Line endings normalized (LF)
- [x] File names are lower-case-with-hyphens
- [x] Preview image uses only synthetic and sanitized data (`assets/preview.png`)

### CODE REVIEW

**Comment by davidkaya** [comment]
Addressed all Copilot review findings in 45b50dd5:

- added a per-instance 256-bit capability token and require it on the HTML, JSON, and SSE routes
- constrained `events.jsonl` resolution to the local `session-state` root and reject escaping paths
- reset the latest-call fields when a selected session has no timeline entries
- included cache-read and cache-write tokens in per-agent totals and the token breakdown
- cancel superseded dashboard requests with `AbortController`
- make the back button explicitly request `currentSessionId`
- regenerated the synthetic preview and required `docs/README.agents.md` output

Validation rerun: `npm start`, `npm run website:data`, `npm run plugin:validate`, extension syntax checks, traversal checks, LF normalization, and `git diff --check` all pass.

**Comment by davidkaya** [comment]
Addressed the follow-up Copilot review in 49ce2276: the embedded bootstrap snapshot now renders immediately, range changes retain the current view while loading, the collapsible heading uses valid heading/button nesting, and both segmented controls expose explicit `role="group"` semantics. The synthetic preview and full validation sequence were refreshed.

**Comment by davidkaya** [comment]
Addressed the third Copilot pass in 95f9a358: current-session totals now prefer live runtime metrics, the responsive back button has a persistent accessible name, and history drill-down no longer forces animated scrolling. The live-total unit check, syntax checks, build, plugin validation, LF normalization, and whitespace checks pass.

**Comment by davidkaya** [comment]
Addressed the fourth Copilot pass in 7fda6aa5: persisted and runtime aggregates now merge monotonically so optional runtime credit fields cannot erase stored totals, reasoning remains visible but is excluded from non-overlapping token totals, and sub-second calls retain millisecond precision. Aggregate edge-case tests, syntax, build, plugin validation, LF, and whitespace checks pass.

File: extensions/usage-insights/stats.mjs [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+        cacheReadTokens: 0,
+        cacheWriteTokens: 0,
+        calls: 0,
+        inputTokens: 0,
+        outputTokens: 0,
+        reasoningTokens: 0,
+        sessions: 1,
+        subagents: 0,
+        totalNanoAiu: number(metrics.totalNanoAiu),
+    };
```
[the remark this answers is not in the document]
Comment (davidkaya) [reply extensions/usage-insights/stats.mjs]: Fixed in 6558a26e. Runtime aggregation now falls back to `totalPremiumRequestCost * 1_000_000_000` when `totalNanoAiu` is absent, matching the existing Token Pac-Man handling and preserving the pre-persistence AI-credit total.

File: extensions/usage-insights/stats.mjs [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+        selectedSessionId,
+        agentMetadata,
+    }) {
+        const dbTotals = this.queryAggregate("WHERE session_id = ?", [selectedSessionId]);
+        const liveWindow =
+            selectedSessionId === currentSessionId && currentRuntime
+                ? runtimeAggregate(currentRuntime)
+                : undefined;
+        const selectedTotals = mergeAggregates(dbTotals, liveWindow);
+        const agents = this.getSessionAgents(selectedSessionId, agentMetadata);
```
[the remark this answers is not in the document]
Comment (davidkaya) [reply extensions/usage-insights/stats.mjs]: Fixed in 6558a26e. Once SQLite has usage rows, every current-session section now uses that single persisted snapshot; runtime totals are used only before the first row is persisted. This keeps totals, agents, models, and timeline internally consistent.

File: extensions/usage-insights/renderer.mjs [hunk 3]
```diff
@@ -0,0 +1,931 @@
...
+        </button>
+        <span class="live" id="liveStatus" aria-live="polite">Live</span>
+      </div>
+    </header>
+
+    <section class="section" aria-labelledby="creditsHeading">
+      <div class="section-head">
+        <h2 id="creditsHeading">
+          <button class="collapsible" id="chartToggle" type="button" aria-expanded="true">
+            <span class="chevron">›</span>AI credits · <span id="chartCredits">—</span>
```
[the remark this answers is not in the document]
Comment (davidkaya) [reply extensions/usage-insights/renderer.mjs]: Fixed in 6558a26e. The decorative chevron now has `aria-hidden="true"`, so the accessible button name contains only the section label and current credit value.

File: extensions/usage-insights/renderer.mjs [hunk 4]
```diff
@@ -0,0 +1,933 @@
...
+      const plotWidth = width - left - right;
+      const plotHeight = height - top - bottom;
+      const total = Math.max(.001, calls.reduce((sum, call) => sum + call.aiCredits, 0));
+      let cumulative = 0;
+      const points = calls.map((call) => {
+        cumulative += call.aiCredits;
+        const x = left + (call.endedAtMs - selected.timeline.startedAtMs) / selected.timeline.durationMs * plotWidth;
+        const y = top + plotHeight - cumulative / total * plotHeight;
+        return { x, y };
+      });
```
[the remark this answers is not in the document]
Comment (davidkaya) [reply extensions/usage-insights/renderer.mjs]: Fixed in 67b96100. The real AI-credit total is retained for labels, while only the plotting denominator uses `Math.max(total, Number.EPSILON)`, so zero-cost and sub-0.001-credit sessions remain accurately labeled.

File: extensions/usage-insights/stats.mjs [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+                .prepare(`
+                    SELECT id, summary, repository, branch, cwd, created_at, updated_at
+                    FROM sessions
+                    WHERE id = ?
+                `)
+                .get(sessionId) || {};
+        const app =
+            this.appDb
+                ?.prepare(`
+                    SELECT title, agent, model, mode, created_at, updated_at
```
[the remark this answers is not in the document]
Comment (davidkaya) [reply extensions/usage-insights/stats.mjs]: Fixed in 0aae213f. Optional `data.db` metadata is now schema-probed and queries only columns that actually exist, so installations without `updated_at`, `title`, `agent`, or `mode` fall back cleanly to `session-store.db`. The same change also normalizes SQLite timestamps for UTC timeline parsing and `datetime(...)` range comparisons.

**Comment by aaronpowell** [comment]
I notice that this is a canvas that is a mirror of code that lives in an external repo. It might be more logical to submit it as an external plugin rather than within the repo directly so that rather than having to copy the code across when you do new iterations, you can bump the version number + SHA/ref for the marketplace.

File: plugins/external.json [hunk 6]
```diff
@@ -1252,6 +1252,30 @@
...
+      "canvas",
+      "copilot-extension",
+      "session-analytics",
+      "token-usage",
+      "usage-insights"
+    ],
+    "source": {
+      "source": "github",
+      "repo": "davidkaya/github-copilot-usage-insights",
+      "sha": "206d53faab1d8a3331de96e203d0206ad21845aa"
```
[the remark this answers is not in the document]
Comment (davidkaya) [reply plugins/external.json]: Fixed in c611de9d. Restored the local Usage Insights extension and matching plugin layout, so this PR no longer points at the un-packaged upstream repository root.

File: plugins/external.json [hunk 7]
```diff
@@ -1252,6 +1252,30 @@
       "sha": "f8abab778cfb5b57a3c745d540631b7aab15d5b3"
     }
   },
+  {
+    "name": "usage-insights",
```
[the remark this answers is not in the document]
Comment (davidkaya) [reply plugins/external.json]: Fixed in c611de9d. Removed the direct `plugins/external.json` listing; this PR now keeps Usage Insights as the bundled plugin and does not bypass the external-plugin intake workflow.

**Comment by davidkaya** [comment]
> I notice that this is a canvas that is a mirror of code that lives in an external repo. It might be more logical to submit it as an external plugin rather than within the repo directly so that rather than having to copy the code across when you do new iterations, you can bump the version number + SHA/ref for the marketplace.

That's a good point, I somehow missed it's possible to do it that way. Created an external plugin intake request: #3005
